### DISCUSSION TITLE
Refetch data on router.replace

### DISCUSSION THREAD

**Discussion by mercore** [vercel/next.js#42397] [Help] [answered]
Hello,
I have a category page on my application where I list all the categories. I use getServerSideProps to fetch the categories. I also have a page where I can create a category. After I submit the category and everything is ok, I just do router.replace() to go back on the categories page. On localhost when the replace is made I can see the new category created but when I build the app and upload it to my cPanel, after I create the category and go back to categories page I can't see the new category there, I have to refresh the page... Does anyone know how to solve this?

This happens on all pages, for example when I login and the user is correct, I make a router.replace("/dashboard"), on cPanel, after login I have to refresh the page to redirect me to /dashboard.
I have to mention that if I build the app and start it on localhost it works fine, only on cPanel I have this problem. Thank you.
On cPanel I have Node 16.17.1.
Sorry for my bad english...

**Comment by icyJoseph** [comment, marked as answer]
__Assuming pages directory__

On CPanel, open the network tab, and clear it, after you create the category and router.replace triggers, can you see a `.json` request (the JSON props for GSSP), being fetched from the server?

- If so, does it contain the data you expect? if not, we have a race condition. Gotta look down into the async process of creating the category, and navigating
- If not, what next.js version do you have? are you using middleware?

## Edit

OP's using a PWA workbox strategy of `stale-while-revalidate` which doesn't play with next-data fetching functions, as they require a page-reload to see update data. `network-first-falling-back-to-cache` is much better for next-data requests.

**Reply by mercore** [reply]
Thank you for your response.
I have two `.json`, one is `categories.json` and the other one has the same name but it has an setting icon in front of it.
That `.json` with the icon in front has the data I expect but it doesn't appear on the page, the other `categories.json` has an empty array.
I have the last next.js version `13.0.0`, I'm not using middleware

**Reply by icyJoseph** [reply]
Alright, so you have a PWA correct? Most PWA's default to https://developer.chrome.com/docs/workbox/modules/workbox-strategies/#stale-while-revalidate which means you need to reload the page to see the new data.

This is obviously bad for GSSP, but before I continue could you confirm the PWA usage?

**Reply by mercore** [reply]
Yes, it is a PWA app.

**Reply by icyJoseph** [reply]
Ok you need to change the workbox strategy for your page to: https://developer.chrome.com/docs/workbox/modules/workbox-strategies/#network-first-network-falling-back-to-cache

What PWA framework are you using? `next-pwa`?

**Reply by icyJoseph** [reply]
You could refer to this answer, https://github.com/vercel/next.js/discussions/39973#discussioncomment-3482901 - In general all of next-data bits should, I think, use `network-first-falling-back-to-cache`.

Reposting for visibility:

```js
/** @type {import('next').NextConfig} */
const withPWA = require("next-pwa");
const runtimeCaching = require("next-pwa/cache");

const nextDataIndex = runtimeCaching.findIndex(
  (entry) => entry.options.cacheName === "next-data"
);

if (nextDataIndex !== -1) {
  runtimeCaching[nextDataIndex].handler = "NetworkFirst";
} else {
  throw new Error("Failed to find next-data object in runtime caching");
}

// now runtimeCaching has been mutated
module.exports = withPWA({
  pwa: {
    dest: "public",
    runtimeCaching,
    register: true,
    skipWaiting: true,
    buildExcludes: [/middleware-manifest.json$/],
  }
});

```

**Reply by mercore** [reply]
Yes, I'm using `next-pwa`.

**Reply by icyJoseph** [reply]
@mercore try with the above suggestion in `next.config.js`

**Reply by icyJoseph** [reply]
Good to purge the old service worker too. Service Workers can be difficult to work with:

https://twitter.com/sanketsahu/status/1133223194888773632

**Reply by mercore** [reply]
**Edit:** Looks like is working now, the data is fetching, I think there was some caching.

Using the config you send in that way throws me this error `'pwa' property is not expected to be here.`
I tried to change it like this, I'm not sure if it is correct. It doesn't throw any error, but still not fetch the data.
```js
/** @type {import('next').NextConfig} */
const runtimeCaching = require("next-pwa/cache");

const nextDataIndex = runtimeCaching.findIndex(
  (entry) => entry.options.cacheName === "next-data"
);

if (nextDataIndex !== -1) {
  runtimeCaching[nextDataIndex].handler = "NetworkFirst";
} else {
  throw new Error("Failed to find next-data object in runtime caching");
}

const withPWA = require("next-pwa")({
  dest: "public",
  runtimeCaching,
  register: true,
  skipWaiting: true,
  buildExcludes: [/middleware-manifest.json$/],
  disable: process.env.NODE_ENV === 'development'
})

module.exports = withPWA({
  reactStrictMode: false
});
```

**Reply by icyJoseph** [reply]
Ah right they updated the next.config object in one of the most recent releases... good stuff though, great contribution to knowledge!
